cluster,kafka: auto-populate cluster_id and prefix it with "redpanda."

jcsp opened the pull request:
## Cover letter

We already generate a unique ID in MetricsReporter.  Use this to initialize `cluster_id` in the cluster configuration if it isn't already set.

This is a slightly obscure place to do the initialization, but there wasn't an obviously better home, and I'm not sure this particulars piece of initialization is big enough to justify getting it's very own class within the controller.

Interesting things about the implementation here:
- I've put a dummy "redpanda.initializing" cluster ID in, for if a Kafka client asks before we've initialized the cluster ID.  I'm interested in feedback from Kowl folks on whether this is helpful or not -- is it better for us to say nothing, or to at least say enough to indicate that we are Redpanda?
- We do need at least two controller log messages before the cluster ID gets initialized.  On many systems this is immediate: the feature manager writes one message right away, and the configuration manager also writes a message if there are any non-default settings at all.  A system with a totally default configuration would have to wait until it saw its first user/topic creation though (that's where the "redpanda.initializing" dummy value would come into play)

## Release notes

### Improvements

* The Kafka cluster ID is now automatically populated with a "redpanda.<uuid>" string if not set by the user.

jcsp commented:
Ping @weeco

weeco commented:
Does that mean that the Redpanda nodes may respond to Kafka requests before the cluster ID is not yet fully initialized and may change?

While we are mostly interested in whether or not we are talking to a Redpanda cluster, the cluster id can also be used as some unique identifier for a certain target cluster (e.g. to store configurations for that specific cluster somewhere). This brings the following challenge with it: How can we know whether a cluster.id is now initialized or not? Can we assume that the "pre-initialized" clusterId `redpanda.initializing` is static and therefore is not suitable as a unique identifier?

jcsp commented:
> Does that mean that the Redpanda nodes may respond to Kafka requests before the cluster ID is not yet fully initialized and may change?

Right.

> Can we assume that the "pre-initialized" clusterId redpanda.initializing is static and therefore is not suitable as a unique identifier?

Yes, with this PR in its current form, Kowl could treat that as a magic value & we could commit to never changing it in redpanda core.

This PR is basically the "lightweight version" of the change, that attaches the cluster ID initialization to a periodic thing that already exists.  A stricter (more work) version of the change would be to generate the cluster ID as early as we possibly can, _and_ refuse to serve any Kafka requests until that initialization has happened.  In that stricter version there would be no redpanda.initializing.

I think the question for right now is: is this slightly sloppy version useful, or should we hold off until we can do the strict version?

senior7515 commented:
@mmaslankaprv would be good to bring this ID into the aggregate metrics layer w/ the json payload.

jcsp commented:
Note to self: don't merge this until fixing config import/export to exclude cluster_id -- that's the downside to cluster_id being stored as a config option, it gets picked up by someone who is trying to replicate the config of one cluster to another, but we don't want people setting up a bunch of clusters with identical IDs.

twmb commented:
What can be done to prevent cluster ID changes during initialization? This basically breaks the immutability of "The cluster_id should be unique,  immutable, auto-generated" as mentioned in KIP-78: https://cwiki.apache.org/confluence/display/KAFKA/KIP-78%3A+Cluster+Id

To me it sounds like cluster_id should be constant, fixed, and not settable by the user. I'm not sure if there's any ecosystem behavior around the fixed aspect, but I'd somewhat guess there were. It also seems like a big footgun to have the ability to change it (whoops changed my ID, now everything is paging because everything used the old ID).

jcsp commented:
> What can be done to prevent cluster ID changes during initialization? This basically breaks the immutability of "The cluster_id should be unique, immutable, auto-generated" as mentioned in KIP-78: https://cwiki.apache.org/confluence/display/KAFKA/KIP-78%3A+Cluster+Id

The changing during initialization is part of making the ID line up with the ID we currently use in the metrics reporter -- we need to initialize it using the same algorithm, which is currently to wait for the second controller log message before doing some hashing to generate the ID.  That second controller log message is usually _but not always_ written out on cluster's first start, in some cases it has to wait for some user action like creating a user/topic.

This is the minimal v1 to get enough in here to support kowl: there should be a v2 of this where we:
- Make the ID intialize as soon as the controller log is ready for us to generate our UUID, rather than on the first metrics reporter tick
- Make the Kafka API listener wait until the ID is ready, so that Kafka clients can never see the transient "redpanda.initializing" state.

That'll be a clean update path -- any tool that copes with the initial few seconds of redpanda.initializing will not need updating when we eventually cut over to a mode where we set the UUID from day 1.

> To me it sounds like cluster_id should be constant, fixed, and not settable by the user. I'm not sure if there's any ecosystem behavior around the fixed aspect, but I'd somewhat guess there were. It also seems like a big footgun to have the ability to change it (whoops changed my ID, now everything is paging because everything used the old ID).

I don't really want the user to change the cluster ID, but I do want to:
- respect existing cluster IDs that have been set externally before we added this feature
- keep the door open for orchestrators/installers to set their own cluster IDs, so that they can e.g. make an ID of a cluster line up with a UUID they had already generated in their database.

twmb commented:
> respect existing cluster IDs that have been set externally before we added this feature
This exists? I didn't know that, and no cluster ID has ever shown up in a metadata response.
> ID of a cluster line up with a UUID they had already generated in their database
Is this important?

My line of questioning is basically: does Confluent allow cluster ID customization in their cloud? The immutability aspect is compelling. If Confluent does not allow customization, I wonder how beneficial this is at all. As well, cluster IDs are used for closest replica fetching -- it seems to me like the ability to change or customize IDs at all is providing users with an unnecessary footgun. Is it detrimental to keep things fixed and internal to start and only loosen restrictions if the need arises from end users?

jcsp commented:
> My line of questioning is basically: does Confluent allow cluster ID customization in their cloud? The immutability aspect is compelling. If Confluent does not allow customization, I wonder how beneficial this is at all. As well, cluster IDs are used for closest replica fetching -- it seems to me like the ability to change or customize IDs at all is providing users with an unnecessary footgun. Is it detrimental to keep things fixed and internal to start and only loosen restrictions if the need arises from end users?

This PR isn't introducing the cluster_id or making it editable: it was already there and was already editable.  This PR is just auto-initializing it so that we can be sure that clusters in the wild do not have a null cluster_id.

I'm not opposed to coming back later to do something stricter -- as we've seen in the related rpk change to import/export, it's just a bit weird to have this as a configuration property at all.

dotnwat commented:
> The Kafka cluster ID is now automatically populated with a "redpanda." string if not set by the user.

afaik kafka always uses a uuid here. so i'm wondering if this format should be considered as a stable part of the api?

EDIT: answering my own question from https://github.com/redpanda-data/redpanda/issues/3874#issuecomment-1078519976 I used `foo_cluster`, so apparently not.

> I've put a dummy "redpanda.initializing" cluster ID in, for if a Kafka client asks before we've initialized the cluster ID.

this one is interesting. so an early query for cluster.id might a different value than a later query? ksql queires for the cluster.id shortly after booting up (simple docker-compose example linked above) and i'm wondering if ksql will notice this change and be upset?

---

Thinking a bit further ahead to rpc.v2+ in which we may have a handshake where we want to do something like verify that a connection is between two nodes in the same cluster, would this be (1) something we could use in that handshake and (2) will it be available early enough in that process? i suppose we could at some point cache the cluster id in a write-once file that can be read on demand to handle the early message case. just thinking through those scenarios.

dotnwat commented:
Will review on Monday! Sorry about the delay here

dotnwat reviewed (commented):
Do I understand correctly that the way this works is as follows:

1. cluster_id in metrics reporter is always generated, but it might take some time for this to actually occur (e.g. two entries in the log, but also replay required on startup).
2. the cluster_id in metrics reporter is polled. as soon as it's noticed to have been generated, it is set in centralized config (if not already set)
3. until cluster_id is propagated to the cluster, a sentinel value is returned to kafka clients that query metadata.

---

The PR generally looks good to me.

It all seems to hinge on whether the early query and changing cluster id is problematic. The real test of early query is probably kSQL docker-compose example. But to understand if a changing cluster id is problematic is much harder as it probably depends entirely on how applications are written.

---

A few thoughts:

1. using the metrics reporter generated cluster id as a default cluster_id seems completely fine, but the reverse does not seem safe--exposing an internal name (e.g. maybe a code name) to the outside world. so this suggests that is fine to continue to let metrics reporter exist in isolation and generate an id however it wants to.
2. presumably there is a convenient bootstrap phase we could piggyback on to generate a uuid and store this in the controller log
3. requiring that a node replay the controller log before booting up doesn't seems entirely crazy (or perhaps only do this when a node is joining the cluster for the first time). but in either case, we could always cache the uuid on disk so it is available at early bootup.

jcsp commented:
>using the metrics reporter generated cluster id as a default cluster_id seems completely fine, but the reverse does not seem safe--exposing an internal name (e.g. maybe a code name) to the outside world. so this suggests that is fine to continue to let metrics reporter exist in isolation and generate an id however it wants to.

This might kind of limit our options with unified support experience across Kowl and redpanda -- I'm imagining if Kowl provided some kind of debug dump functionality, it would be good if it used the same cluster ID that redpanda uses for metrics reports.

However, we're in a kind of weird in-between place where this is a config property, so users can in principle change it anyway, we're just making them identical by default.

> presumably there is a convenient bootstrap phase we could piggyback on to generate a uuid and store this in the controller log

Hmm, sorta, I guess we have the place where the original node adds itself to the members_table.  That and the feature manager initial message.

> requiring that a node replay the controller log before booting up doesn't seems entirely crazy (or perhaps only do this when a node is joining the cluster for the first time). but in either case, we could always cache the uuid on disk so it is available at early bootup.

Yeah.  Currently, the cluster_id in the config is already stored on local disk so available from the very beginning of startup, so as/when we refine any of this, our options are open.

Other than the leg-work to refactor ID generation to some earlier point, it feels like the key behavioral decision here is whether we continue to allow setting the ID externally, or whether we always bake in a UUID.

twmb commented:
> allow setting the ID externally, or whether we always bake in a UUID

twmb commented:
> allow setting the ID externally, or whether we always bake in a UUID
fwiw my vote is baking in the UUID, and as Noah mentioned, we don't know what in the ecosystem relies on a static ID :D

jcsp commented:
/backport v22.1.x

kpweiler commented:
FWIW - I installed a fresh dev cluster from the apt repos at `22.1.3` and this wasn't populated:
```
# rpk cluster config get cluster_id
null
```
NBD to just set it manually though.

jcsp commented:
@kpweiler this PR merged into the `dev` branch of redpanda after 22.1.x was branched.  The backport to 22.1.x made it into version 22.1.4 I believe.

weeco commented:
I can confirm that this is set in v22.1.4
